### ISSUE TITLE
TopK is at best 30 times slower in ONNX than it is in Pytorch

### ISSUE THREAD

**Issue by d-colwell** [microsoft/onnxruntime#1830]
**Describe the bug**
The TopK implementation in ONNX is incredibly slow. The operation performed in Pytorch/Tensorflow will complete in ~30ms, in ONNX will take 1130ms

**Urgency**
This is significantly impacting our ability to deploy models requiring multiple output classifications to production.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10, C#, latest drivers and updates
- ONNX Runtime installed from (source or binary): NuGet package
- ONNX Runtime version: 0.5
- Python version: 3.7.3 (Pytorch version 1.2)
- Visual Studio version (if applicable): VS2019
- GCC/Compiler version (if compiling from source): Not compiled from source
- CUDA/cuDNN version: CUDA 10, CUDNN 7.x
- GPU model and memory: NVidia Titan RTX - 24GB Memory

**To Reproduce**
Here is the code to export the ONNX model from Pytorch

    class TopkExample(nn.Module):
        def __init__(self):
            super(TopkExample, self).__init__()
            self.expander = nn.Conv2d(3,192,1,1)

        def forward(self,x:torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
            x = self.expander(x)
            x_val, x_indices = torch.topk(x,3,dim=1)
            return (x_val,x_indices)

    model = TopkExample()
    model = model.cuda()
    sample_input = torch.rand(1,3,544,1920).cuda()
    s = datetime.now()
    max_mem = 0
    for i in range(10):
        out = model(sample_input)
        mem = torch.cuda.max_memory_allocated()
        if mem > max_mem:
            max_mem = mem
        torch.cuda.synchronize()
    f = datetime.now()
    print(f'{(f-s).microseconds//1000}ms for 10 iterations, max memory = {max_mem//1024//1024}MB')

    model.eval()
    torch.onnx.export(model,sample_input,'topk_test.onnx',input_names=['img'],output_names=['values','indices'])

This is the output:
[topk_test(actually onnx file renamed for Github).txt](https://github.com/microsoft/onnxruntime/files/3607967/topk_test.actually.onnx.file.renamed.for.Github.txt)

Here is the code i used to test the ONNX values

       static void Main(string[] args)
        {
            var sess = new InferenceSession(@"F:\Code\Repos\APAC\VisualTextRecognition\topk_test.onnx", SessionOptions.MakeSessionOptionWithCudaProvider());
            foreach (var item in sess.InputMetadata)
            {
                Console.WriteLine($"{item.Key} - [{item.Value.Dimensions.Select(x => x.ToString()).Aggregate((x, y) => $"{x},{y}")}]");
            }

            foreach (var item in sess.OutputMetadata)
            {
                Console.WriteLine($"{item.Key} - [{item.Value.Dimensions.Select(x => x.ToString()).Aggregate((x, y) => $"{x},{y}")}]");
            }
            long[] milliseconds = new long[10];
            Tensor<float> img = new DenseTensor<float>(new int[] { 1, 3, 544, 1920 });
            var named_img = new NamedOnnxValue[] { NamedOnnxValue.CreateFromTensor("img", img) };
            for (int i = 0; i < 10; i++)
            {
                var start = DateTime.Now;
                var input = named_img;
                using (var results = sess.Run(input))
                {
                    foreach (var result in results)
                    {
                        var tensor = result.AsTensor<float>();
                        var data = tensor?.ToArray();
                    }
                }
                var end = DateTime.Now;
                milliseconds[i] = (long)(end - start).TotalMilliseconds;
                Console.WriteLine($"{(end - start).TotalMilliseconds}ms");
            }
        }

**Expected behavior**
I expected this to perform quite similarly in ONNX and in Pytorch, the time difference is significant

**Other things we tried**
Using Max is a lot faster, but we need more than one return value.

**Comment by ybrnathan** [comment]
@d-colwell , thank you for reporting this issue. Adding TopK Cuda support is on our list, this operator is also used by Mask_rcnn and Fast_rcnn model. Cuda support will be added in next ONNX Runtime release.

**Comment by d-colwell** [comment]
Thanks @ybrnathan,
Any idea roughly when the next release is coming out? I'm also hopeful that it will bump the power version to 10 :)

**Comment by jinfagang** [comment]
@ybrnathan  Hopefully there could be an example to inference maskrcnn or fastercnn with ONNXRuntime GPU or GPU fusion with TensorRT. I am still not able to inference through onnxruntime TensorRT and GPU on maskrcnn onnx model.
What would be the release time around BTW?

**Comment by jinfagang** [comment]
@ybrnathan  Does it supported now?

**Comment by ybrnathan** [comment]
There is a PR in progress (#1919) as pointed by Hari. There are still a few operators that required by Mask-RCNN which still falls back to CPU, and the Cuda kernel for those Ops are planned work (current ETA is end of Oct)

**Comment by jinfagang** [comment]
@ybrnathan  Glad to here that, but, I haven't able to run maskrcnn on onnxruntime GPU at all (they do not support dynamic input). Have u able to run maskrcnn on onnxruntime mostly in TensorRT? How much latency do u got?

**Comment by ybrnathan** [comment]
I am able to run mask-rcnn model zoo model with its test data. It's slow, but we are able to get to around 120ms in a private branch. Those changes will be merged to master by end of Oct.

**Comment by jinfagang** [comment]
120ms still slow.
The reason for slow speed is most operators still runs on CPU.

**Comment by ybrnathan** [comment]
#1919 has merged to master.
